### PR TITLE
(CLOUD_1505) validate functions

### PR DESCRIPTION
**Pull request by sheenaajay**
Request to review the changes for replacing validate functions with datatypes and assert_type function.
Following jobs are running clean
bundle exec rake test
bundle exec rake spec
bundle exec rake acceptance

### CODE REVIEW

File: manifests/registry.pp [hunk 1]
```diff
@@ -73,7 +73,7 @@
     $local_user_strip = regsubst($local_user, '-', '', 'G')
 
     $_pass_hash = $pass_hash ? {
-      Undef   => pw_hash("${title}${auth_environment}${auth_cmd}${local_user}", 'SHA-512', $local_user_strip),
+      'Undef'   => pw_hash("${title}${auth_environment}${auth_cmd}${local_user}", 'SHA-512', $local_user_strip),
```
Comment (davejrt) [inline manifests/registry.pp]: @sheenaajay have you run this through travis without the qoutes? The PR to remove them was valid and needs to as is.

File: manifests/init.pp [hunk 2]
```diff
@@ -357,163 +357,184 @@
...
-  validate_array($daemon_environment_files)
-  validate_array($log_opt)
-  validate_bool($tls_enable)
-  validate_bool($ip_forward)
-  validate_bool($iptables)
-  validate_bool($ip_masq)
-  if $icc != undef {
-    validate_bool($icc)
+  if $version {
+    assert_type(String[1], $version)
```
Comment (HelenCampbell) [inline manifests/init.pp]: Not 100% sure, but I think you can put this check in the type declaration itself, like - Optional[String[1]] $version - Will save you the extra couple of lines of the assert type here :)

File: manifests/init.pp [hunk 3]
```diff
@@ -357,163 +357,184 @@
...
-  validate_string($fixed_cidr)
-  validate_string($default_gateway)
-  validate_string($bip)
 
+  if $::osfamily {
+    assert_type(Pattern[/^(Debian|RedHat|Archlinux|Gentoo)$/], $::osfamily) |$a, $b| {
+      fail('This module only works on Debian or Red Hat based systems or on Archlinux as on Gentoo.')
+    }
+  }
+  assert_type(Boolean, $manage_kernel)
```
Comment (HelenCampbell) [inline manifests/init.pp]: When you define the params at the top with the Data Type it's pretty much the same as doing an assert type on the param - If it comes through as anything other than the specified Data Type it'd fail further up at the param definition, so you can take all these assert types out!

File: manifests/init.pp [hunk 4]
```diff
@@ -541,15 +562,19 @@
     if(!$tcp_bind) {
         fail('You need to provide tcp bind parameter for TLS.')
     }
-    validate_string($tls_cacert)
-    validate_string($tls_cert)
-    validate_string($tls_key)
+    assert_type(String[1], $tls_cacert)
```
Comment (HelenCampbell) [inline manifests/init.pp]: Again, if it makes sense, I'd remove these assert types and add the string lower limit to the data type - Optional[String[1]] $tls_cacert

File: manifests/registry.pp [hunk 5]
```diff
@@ -33,19 +33,19 @@
...
+  Optional[String] $pass_hash   = undef,
+  Optional[String] $email       = undef,
+  Optional[String] $local_user  = 'root',
+  Optional[String] $version     = $docker::version,
+  Optional[Boolean] $receipt     = true,
 ) {
   include docker::params
 
-  validate_re($ensure, '^(present|absent)$')
+  assert_type(Pattern[/^present$|^absent$/], $ensure)
```
Comment (HelenCampbell) [inline manifests/registry.pp]: You can use Pattern[] as a data type! So you could remove this line and have the definition further up as -
 Optional[Pattern[/^present$|^absent$/]] $ensure - https://puppet.com/docs/puppet/5.3/lang_data_abstract.html#pattern

File: manifests/run.pp [hunk 6]
```diff
@@ -120,39 +120,53 @@
   $service_name = $docker::service_name
   $docker_group = $docker::docker_group
 
-  validate_re($image, '^[\S]*$')
-  validate_re($title, '^[\S]*$')
-  validate_re($memory_limit, '^[\d]*(b|k|m|g)$')
-  validate_re($ensure, '^(present|absent)')
+  if $image {
+  assert_type(Pattern[/^[\S]*$/], $image)
```
Comment (HelenCampbell) [inline manifests/run.pp]: As above - I'd consider moving the params here to data type Pattern to be able to remove the assert_types here.

**Comment by sheenaajay** [comment]
@davejrt Thanks alot Dave.Will check the local environment and reexecute the unit testcases.Travis jobs are running clean.

**Comment by sheenaajay** [comment]
@HelenCampbell Thanks alot for the review.

File: manifests/compose.pp [hunk 7]
```diff
@@ -21,17 +21,11 @@
...
 #
 class docker::compose(
-  $ensure = 'present',
-  $version = $docker::params::compose_version,
-  $install_path = $docker::params::compose_install_path,
-  $proxy = undef
+  Optional[Pattern[/^present$|^absent$/]] $ensure                                                                         = 'present',
+  Optional[String] $version                                                                                               = $docker::params::compose_version,
+  Optional[String] $install_path                                                                                          = $docker::params::compose_install_path,
+  Optional[Pattern[/^(https?:\/\/)?([^:^@]+:[^:^@]+@|)([\da-z\.-]+)\.([\da-z\.]{2,6})(:[\d])?([\/\w \.-]*)*\/?$/]] $proxy = undef
```
Comment (jeefberkey) [inline manifests/compose.pp]: I have to scroll to see what the parameter names are, maybe a custom type alias is a good idea here

**Comment by sheenaajay** [comment]
@davejrt @HelenCampbell Thanks for the reviews.All comments are incorporated.

File: manifests/compose.pp (hunk 7 above)
Comment (jeefberkey) [reply manifests/compose.pp]: Same comment for every large data type

Comment (sheenaajay) [reply manifests/compose.pp]: @jeefberkey Thank you for the valuable comment

Comment (sheenaajay) [reply manifests/compose.pp]: @jeefberkey The large data types are not repeating alot to create a data type.So will specify the assert_type function for all large data types outside the variable declaration area so keep it short.

**Comment by davejrt** [comment]
LGTM

**Comment by esalberg** [comment]
We are having issues upgrading past this commit as well due to dm_use_deferred_deletion and dm_use_deferred_removal being set to String instead of Boolean.  Perhaps it was set to String because the default was set to 'undef' (converted to false)?  My concern is that while the String update breaks our config, I don't know if other installations might be affected the other way (although lint doesn't like a quoted true/false, so I'm assuming most probably used booleans by default).

**Pull request merged**